### PR TITLE
Switch to kotlin-compile-testing for JsonClassCodegenProcessorTest

### PR DESCRIPTION
**Pull request by ZacSweers**
This switches to use [kotlin-compile-testing](https://github.com/tschuchortdev/kotlin-compile-testing), a compile-testing-inspired library for Kotlin. Aside from allowing us to not have to maintain this ourselves, it's also much more feature complete. Probably best summed up by the fact that we can now actually [write tests that expect successful compilations](https://github.com/square/moshi/commit/0e08d5b2cafaec440076c0b2c3fbe96dfb740bad).

### CODE REVIEW

**Comment by ZacSweers** [comment]
The CI failure is strange, because it doesn't appear to actually match the code changes of the PR?

```
[ERROR] Failed to execute goal org.jetbrains.kotlin:kotlin-maven-plugin:1.3.40:test-compile (test-compile) on project moshi-kotlin-codegen: Compilation failure: Compilation failure: 
[ERROR] /home/travis/build/square/moshi/kotlin/codegen/src/test/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessorTest.kt:[120,16] Unresolved reference: KotlinCompilerCall
[ERROR] /home/travis/build/square/moshi/kotlin/codegen/src/test/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessorTest.kt:[122,21] Unresolved reference: Processor
[ERROR] /home/travis/build/square/moshi/kotlin/codegen/src/test/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessorTest.kt:[131,43] Unresolved reference: ExitCode
```

File: kotlin/codegen/pom.xml [hunk 1]
```diff
@@ -56,32 +56,31 @@
       <artifactId>assertj-core</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>com.google.truth</groupId>
+      <artifactId>truth</artifactId>
+      <version>1.0</version>
```
Comment (JakeWharton) [inline kotlin/codegen/pom.xml]: versions should go in `<dependencyManagement>` section of root pom.

File: kotlin/codegen/pom.xml [hunk 2]
```diff
@@ -56,32 +56,31 @@
...
-      <scope>test</scope>
+      <groupId>me.eugeniomarletti.kotlin.metadata</groupId>
+      <artifactId>kotlin-metadata</artifactId>
     </dependency>
     <dependency>
-      <groupId>org.jetbrains.kotlin</groupId>
-      <artifactId>kotlin-annotation-processing-embeddable</artifactId>
+      <groupId>com.github.tschuchortdev</groupId>
+      <artifactId>kotlin-compile-testing</artifactId>
+      <version>1.2.2</version>
```
Comment (JakeWharton) [inline kotlin/codegen/pom.xml]: versions should go in `<dependencyManagement>` section of root pom.

File: kotlin/codegen/pom.xml [hunk 3]
```diff
@@ -56,32 +56,31 @@
...
-    <!--
-      Though we don't use compile-testing, including it is a convenient way to get tools.jar on the
-      classpath. This dependency is required by kapt3.
-    -->
-    <dependency>
-      <groupId>com.google.testing.compile</groupId>
-      <artifactId>compile-testing</artifactId>
+      <groupId>com.squareup.okio</groupId>
+      <artifactId>okio</artifactId>
+      <version>2.1.0</version>
```
Comment (JakeWharton) [inline kotlin/codegen/pom.xml]: versions should go in `<dependencyManagement>` section of root pom.

File: kotlin/codegen/src/test/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessorTest.kt [hunk 4]
```diff
@@ -15,367 +15,338 @@
...
+          
+          @JsonClass(generateAdapter = true)
+          class PrivateConstructor private constructor(var a: Int, var b: Int) {
+            fun a() = a
+            fun b() = b
+            companion object {
+              fun newInstance(a: Int, b: Int) = PrivateConstructor(a, b)
+            }
+          }
+          """
```
Comment (JakeWharton) [inline kotlin/codegen/src/test/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessorTest.kt]: why the change from a margin and trimMargin? I think I prefer the old string

Comment (ZacSweers) [reply kotlin/codegen/src/test/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessorTest.kt]: trimMargin breaks syntax highlighting, [which the `kotlin()` factory enables](https://github.com/tschuchortdev/kotlin-compile-testing/pull/13).

image

image

image

File: kotlin/codegen/pom.xml (hunk 3 above)
Comment (ZacSweers) [reply kotlin/codegen/pom.xml]: Part of a4c55f7

File: kotlin/codegen/pom.xml (hunk 2 above)
Comment (ZacSweers) [reply kotlin/codegen/pom.xml]: Part of a4c55f7

File: kotlin/codegen/pom.xml (hunk 1 above)
Comment (ZacSweers) [reply kotlin/codegen/pom.xml]: Part of a4c55f7
